src/EventStore.Core/Services/Storage/StorageWriterService.cs

```diff
@@ -236,7 +236,12 @@ public class StorageWriterService<TStreamId> : IHandle<SystemMessage.SystemInit>
 				_vnodeState != VNodeState.PreReadOnlyReplica)
 				throw new Exception(string.Format("{0} appeared in {1} state.", message.GetType().Name, _vnodeState));
 
-			if (Writer.Checkpoint.Read() != Writer.Checkpoint.ReadNonFlushed()) {
+			if (Writer.CommittedPosition != Writer.NextRecordPosition)
+				throw new Exception("Invariant violation: " +
```

Comment:
probably an invalidoperationexception

i'd be tempted to move this check into the Writer itself and add a hint about why they might be out of sync 'There may be an open transaction' or something like that

Reply:
fixed: added a method named: `HasOpenTransaction()` to `ITransactionFileWriter`